### ISSUE TITLE
Hierarchy enabled in CenterPoint.

### ISSUE THREAD

**Issue by synsin0** [neeharperi/LT3D#6]
I see you disabled the hierarchy settings in all four CenterPoint configs. HIERARCHY={"TRAIN": False ....}. How to enable it during training and inference? Which state is correct for CenterPoint settings?

**Comment by neeharperi** [comment]
Hi @synsin0, 
All CenterPoint configs have the hierarchical loss described in our paper turned on by default (e.g. every "stroller" is also detected as a "pedestrian and "object). The line you're referring to is for the hierarchical softmax ablations we ran. 

I'll provide a bit more detail to give you a better understanding of the config. As an example, I'll refer to [this config](https://github.com/neeharperi/LT3D/blob/main/configs/centerpoint/centerpoint_0075voxel_second_secfpn_dcn_4x8_cyclic_50m_wide_hierarchy_tta_20e_nus.py). To turn off our hierarchical loss, you'll need to change:

- L39: total_class_names = class_names
- L59: nms = ["W", "W", "W", "W", "W", "W", "W", "W", "W", "W", 
       "W", "W", "W", "W", "W", "W", "W", "W"]
- L64: hierarchical_softmax = []
- Comment out L69

The lines you're referring to (L64-72) allow you to define arbitrary groupings for hierarchical softmax. For example, the group ['car', 'vehicle', 'object'] enforces that P(car) = P(car) * P(vehicle) * P(object), where P(car) is the confidence that the detected object is a car. The TRAIN flag on L70 enforces this hierarchical softmax constraint during training, the TEST flag enforces this constraint during inference. by default, we do not perform hierarchical softmax (hence why both flags are False). 

Please let me know if you have any further questions.

**Comment by synsin0** [comment]
I try to understand. For each category, the semantic hierarchy is to detect multi-class, e.g. 'stroller', 'pedestrian', 'object' at the same time. In L601-L611 in centerpoint_head.py, when you set both flags as False, the hierarchical softmax is disabled in both processes. So hierarchical_softmax = [] applies to all default cases. Then where is exactly the hierarchical loss? loss_cls is a common gaussian focal loss, does it applies to multi-class classification?

**Comment by neeharperi** [comment]
The loss_cls is a sigmoid loss, so the classification is not mutually exclusive. We enforce the hierarchical loss by modifying the data loader [here](https://github.com/neeharperi/LT3D/blob/ad8b73dfac054ff1807d74f2f5097a801f848692/mmdet3d/datasets/pipelines/transforms_3d.py#L544) by duplicating each detection and assigning a class label for each level of the hierarchy.

Concretely, nuScenes-LT3D has 18 classes. We add 4 additional classes representing different levels of the hierarchy (e.g. vehicle, pedestrian, movable, object). We add these additional classes (Note: a "stroller" box will have an overlapping  "pedestrian" box and "object" box on top)  and train a 22-class detector. At test-time, we simply ignore the predictions from the vehicle, pedestrian, and object superclasses.

**Comment by synsin0** [comment]
Get! Thanks for your great work and timely instructions!
